Mathlib/Topology/Category/Profinite/Nobeling.lean

```diff
@@ -0,0 +1,1556 @@
+    (π C (ord I · < o')) (isClosed_proj _ _ hC) (contained_proj _ _))
+
+theorem GoodProducts.linearIndependentAux (μ : Ordinal) : P I μ := by
+  refine Ordinal.limitRecOn μ P0 (fun o h ho C hC hsC ↦ ?_)
+      (fun o ho h ↦ (GoodProducts.Plimit o ho (fun o' ho' ↦ (h o' ho'))))
+  have ho' : o < Ordinal.type (·<· : I → I → Prop) :=
+    lt_of_lt_of_le (Order.lt_succ _) ho
+  rw [linearIndependent_iff_sum C hsC ho']
+  refine ModuleCat.linearIndependent_leftExact ?_ ?_ (succ_mono C o) (succ_exact C hC hsC ho')
+      (square_commutes C ho')
+  · exact h (le_of_lt ho') (π C (ord I · < o)) (isClosed_proj C o hC) (contained_proj C o)
+  · exact linearIndependent_comp_of_eval C hC hsC ho' (span (π C (ord I · < o))
+      (isClosed_proj C o hC)) (h (le_of_lt ho') (C' C ho') (isClosed_C' C hC ho')
+      (contained_C' C ho'))
+
+theorem GoodProducts.linearIndependent (hC : IsClosed C) :
+    LinearIndependent ℤ (GoodProducts.eval C) :=
+  GoodProducts.linearIndependentAux (Ordinal.type (·<· : I → I → Prop)) (le_refl _)
+    C hC (fun _ _ _ _ ↦ Ordinal.typein_lt_type _ _)
+
+/-- `GoodProducts C` as a `ℤ`-basis for `LocallyConstant C ℤ`.  -/
+noncomputable
+def GoodProducts.Basis (hC : IsClosed C) :
+    Basis (GoodProducts C) ℤ (LocallyConstant C ℤ) :=
+  Basis.mk (GoodProducts.linearIndependent C hC) (GoodProducts.span C hC)
+
+end Induction
+
+variable {S : Profinite} {ι : S → I → Bool} (hι : ClosedEmbedding ι)
+
+theorem Nobeling : Module.Free ℤ (LocallyConstant S ℤ) := Module.Free.of_equiv'
+  (Module.Free.of_basis <| GoodProducts.Basis _ hι.closed_range) (LocallyConstant.congrLeftₗ ℤ
+  (Homeomorph.ofEmbedding ι hι.toEmbedding)).symm
+
+end NobelingProof
+
+variable (S : Profinite.{u})
+
+open Classical in
+/-- The embedding `S → (I → Bool)` where `I` is the set of clopens of `S`. -/
+noncomputable
+def Nobeling.ι : S → ({C : Set S // IsClopen C} → Bool) := fun s C => decide (s ∈ C.1)
+
+open Classical in
+theorem Nobeling.embedding : ClosedEmbedding (Nobeling.ι S) := by
+  apply Continuous.closedEmbedding
+  · dsimp [ι]
+    refine continuous_pi ?_
+    intro C
+    rw [← IsLocallyConstant.iff_continuous]
+    refine ((IsLocallyConstant.tfae _).out 0 3).mpr ?_
+    rintro ⟨⟩
+    · refine IsClopen.isOpen (isClopen_compl_iff.mp ?_)
+      convert C.2
+      ext x
+      simp only [Set.mem_compl_iff, Set.mem_preimage, Set.mem_singleton_iff,
+        decide_eq_false_iff_not, not_not]
+    · refine IsClopen.isOpen ?_
+      convert C.2
+      ext x
+      simp only [Set.mem_preimage, Set.mem_singleton_iff, decide_eq_true_eq]
+  · intro a b h
+    by_contra hn
+    obtain ⟨C, hC, hh⟩ := exists_clopen_of_totally_separated hn
+    apply hh.2 ∘ of_decide_eq_true
+    dsimp [ι] at h
+    rw [← congr_fun h ⟨C, hC⟩]
+    exact decide_eq_true hh.1
+
+/-- Nöbeling's theorem: the `ℤ`-module `LocallyConstant S ℤ` is free for every `S : Profinite` -/
+theorem Nobeling : Module.Free ℤ (LocallyConstant S ℤ) :=
```

Comment:
I don't think this should go in the root namespace? Why not put this in your namespace, call it Nobeling (although do we have a convention that Nobeling should be the name of the theorem
statement not the proof?) and the other one that nobody should ever use Nobeling_aux.

Reply:
Isn't `Module.Free` a class? Then this could be an instance, right?
I would prefer to call it `LocallyConstant.freeOfProfinite`, and only mention Nobeling in the docstring.